### ISSUE TITLE
IMAP Server responses with escaped quotes are incorrectly handled

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3659]
|<img alt="ali's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| ali reported|
|-|-|
|Trac ID|trac#3659|
|Type|defect|
|Created|2009-02-16 03:42:06Z|

This happens in two places, first the escape is ignored, and then the string is split on the ignored escaped quote.

So I have added a patch which:
   1. tests that fail with the current behaviour
   2. fix for initial parsing to preserve escaped quotes
   3. for quote-splitting which allows for and removes escapes.

Attachments:

* [imap4-parse-bug.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ed5/ed5d0b0e58c5d7b0c9f6dc48d26af7a354af4185/194b589e2aa65481f6258b6b3276c34d22403e64.diff) (3653 bytes) - added by ali on 2009-02-16 03:42:28Z - 
* [imap4-parse-bug2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ed5/ed5d0b0e58c5d7b0c9f6dc48d26af7a354af4185/48be3093d22bc129b1a8605e18de31bac0a91c07.diff) (4307 bytes) - added by ali on 2009-02-23 00:54:59Z - 
* [imap4-parse-bug3.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ed5/ed5d0b0e58c5d7b0c9f6dc48d26af7a354af4185/22e721bac803e11aad5f74fac034e3d5f0ae45d2.diff) (4730 bytes) - added by ali on 2009-03-02 02:42:32Z - 
* [imap4-parse-bug3.2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ed5/ed5d0b0e58c5d7b0c9f6dc48d26af7a354af4185/14cf9a77dca4393df0897b794edc4c7f2160eb76.diff) (4786 bytes) - added by ali on 2009-03-02 02:45:19Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3659 3659
type__defect defect
reporter__ali ali
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__ 
time__1234755726000000 1234755726000000
changetime__1236044756000000 1236044756000000
version__None None
owner__ 
cc__aafshar@...
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|z3p set owner to ali|
|-|-|

1.There aren't any tests for parseNestedParen() handling escapes of non-quote characters.
 1. It doesn't look like actually splitting a quoted strings (having two separate strings in the list) is tested either.
 1. The in* variables in parseNestedParen() and splitOn() should start off as bools, not ints.
 1. There are some pyflakes errors in those files.

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|z3p commented|
|-|-|

Oh, and update the copyright on both files to 2009.

**Comment by twisted-trac** [comment]
|<img alt="ali's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|ali commented|
|-|-|

Thanks for the review. I have attached a new patch.

1. added test case (I think)
2. added test case
3. I don't understand, the original patch changed ints to bools, do you mean vice-versa?
4. They are all old pyflakes failures -> I will create a new ticket.
5. Done.

**Comment by twisted-trac** [comment]
|<img alt="ali's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|ali set owner to z3p|
|-|-|

[#3669](https://github.com/twisted/twisted/issues/3669) for the pyflakes errors

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|z3p removed owner|
|-|-|

I can't be sure I'll review this, so I'm putting it up for grabs.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to ali|
|-|-|

  1. There's still untested changes.  If you do TDD, this won't come up as often. ;)  (On the other hand, it's possible this code wasn't entirely covered in trunk either, I'm not going to bother to check). In imap4.py, at around line 3939, one line was removed and replaced with a conditional.  The tests exercise the positive path of the conditional but not the negative path.  So, for `parseNestedParens`, there's no test for a \ followed by a non-quote inside a quoted string.
  1. The `test_imap.py` copyright statement is still out of date.

**Comment by twisted-trac** [comment]
|<img alt="ali's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|ali commented|
|-|-|

Well, I think I made better tests (tried to be more methodical), and unbroke the code a bit more. And I fixed the copyright. The rfc really only talks about \ and " in quoted strings as being special characters and requiring prefixing with \.

**Comment by twisted-trac** [comment]
|<img alt="ali's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|ali set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

Okay, looks good.  Thanks.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set status to `closed`|
|-|-|

(In [26331]) Apply imap4-parse-bug3.2.diff fixing IMAP4 quoting and escaping parsing

Author: ali
Reviewer: z3p, exarkun
Fixes: [#3659](https://github.com/twisted/twisted/issues/3659)

Fix bugs in two of the low-level IMAP4 parsing functions which caused them to
mis-parse escape sequences inside quoted strings.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
